File: spring-webflux/webflux/r2dbc-coroutine/src/main/java/com/example/webflux/r2dbc/controller/UserController.java
package com.example.webflux.r2dbc.controller;

import com.example.webflux.r2dbc.controller.dto.SignupUserRequest;
import com.example.webflux.r2dbc.controller.dto.UserResponse;
import com.example.webflux.r2dbc.service.UserService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.context.ReactiveSecurityContextHolder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Mono;

@Slf4j
@RestController
@RequestMapping("/api/users")
@RequiredArgsConstructor
public class UserController {

    private final UserService userService;

    @GetMapping("/{userId}")
    public Mono<UserResponse> getUserById(
            @PathVariable String userId
    ) {
        return ReactiveSecurityContextHolder.getContext()
                .flatMap(context -> {
                    String name = context.getAuthentication().getName();

                    if (!name.equals(userId)) {
                        return Mono.error(new ResponseStatusException(HttpStatus.UNAUTHORIZED));
                    }

                    return userService.findById(userId)
                            .map(UserResponse::of)
                            .switchIfEmpty(
                                    Mono.error(new ResponseStatusException(HttpStatus.NOT_FOUND))
                            );
                });
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/signup")
    public Mono<UserResponse> signup(
            @RequestBody SignupUserRequest request
    ) {
        return userService.createUser(
                        request.getName(), request.getAge(), request.getPassword(), request.getProfileImageId()
                )
                .map(UserResponse::of);
    }
}

File: spring-webflux/webflux/r2dbc-coroutine/src/main/kotlin/com/example/webflux/r2dbc/controller/UserController.kt
package com.example.webflux.r2dbc.controller

import com.example.webflux.r2dbc.common.User
import com.example.webflux.r2dbc.controller.dto.SignupUserRequest
import com.example.webflux.r2dbc.controller.dto.UserResponse
import com.example.webflux.r2dbc.service.UserCoroutineService
import org.springframework.http.HttpStatus
import org.springframework.security.core.context.ReactiveSecurityContextHolder
import org.springframework.security.core.context.SecurityContext
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException
import reactor.core.publisher.Mono

@RequestMapping("/api/users")
@RestController
class UserController(
    private val userService: UserCoroutineService,
) {

    @GetMapping("/{userId}")
    fun getUserById(
        @PathVariable userId: String
    ): Mono<UserResponse> {
        return ReactiveSecurityContextHolder.getContext()
            .flatMap { context: SecurityContext ->
                val name = context.authentication.name
                if (name != userId) {
                    return@flatMap Mono.error<UserResponse>(
                        ResponseStatusException(HttpStatus.UNAUTHORIZED)
                    )
                }
                userService.findById(userId)
                    .map { user: User -> UserResponse.of(user) }
                    .switchIfEmpty(Mono.error(ResponseStatusException(HttpStatus.NOT_FOUND)))
            }
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/signup")
    fun signup(
        @RequestBody request: SignupUserRequest
    ): Mono<UserResponse> {
        return userService.createUser(request.name, request.age, request.password, request.profileImageId)
            .map { user: User -> UserResponse.of(user) }
    }
}